bot/cogs/moderation/infractions.py

```diff
@@ -0,0 +1,525 @@
+import logging
+import textwrap
+import typing as t
+
+import dateutil.parser
+import discord
+from discord import Member
+from discord.ext import commands
+from discord.ext.commands import Context, command
+
+from bot import constants
+from bot.api import ResponseCodeError
+from bot.constants import Colours, Event
+from bot.converters import Duration
+from bot.decorators import respect_role_hierarchy
+from bot.utils import time
+from bot.utils.checks import with_role_check
+from bot.utils.scheduling import Scheduler
+from . import utils
+from .modlog import ModLog
+from .utils import MemberObject
+
+log = logging.getLogger(__name__)
+
+MemberConverter = t.Union[utils.UserTypes, utils.proxy_user]
+
+
+class Infractions(Scheduler, commands.Cog):
+    """Server moderation tools."""
+
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+        self.category = "Moderation"
+        self._muted_role = discord.Object(constants.Roles.muted)
+        super().__init__()
+
+    @property
+    def mod_log(self) -> ModLog:
+        """Get currently loaded ModLog cog instance."""
+        return self.bot.get_cog("ModLog")
+
+    @commands.Cog.listener()
+    async def on_ready(self) -> None:
+        """Schedule expiration for previous infractions."""
+        infractions = await self.bot.api_client.get(
+            'bot/infractions',
+            params={'active': 'true'}
+        )
+        for infraction in infractions:
+            if infraction["expires_at"] is not None:
+                self.schedule_task(self.bot.loop, infraction["id"], infraction)
```

Comment:
@scragly made a good point yesterday in #dev-core and I think it needs to be addressed. Because we're using the `on_ready` event to reschedule the expiration of all active infractions with an expiration date and the `on_ready` event does not fire when you reload the cog, reloading the `moderation` extension will erase all of the scheduled tasks:

![2019-10-03_08-48](https://user-images.githubusercontent.com/33516116/66105010-8fc15280-e5ba-11e9-9303-66fa17a54aaa.png)

This is probably a problem that has always existed and has always gone unnoticed, but now is good time to fix it, I think. The solution to this would be to move this rescheduling logic to a regular coroutine that's added to the event loop as a Task at the end of the Infraction cog's `__init__`. I think we only need to add an `await bot.wait_until_ready()` as the first line, but I don't think the rest of the code needs to be changed.

Reply:
Is efd8a30 what you had in mind? Tried it out and it solves the issue. I looked at watchchannels and that assigns the task to a private attribute which is never referenced anywhere. Is that necessary to do here too?